### ISSUE TITLE
[QST] ABI mismatch note on SM90 / H100 compilation

### ISSUE THREAD

**Issue by aakhundov** [NVIDIA/cutlass#906]
When compiling SM90 kernels on an H100 host (e.g., the [`48_hopper_warp_specialized_gemm.cu`](https://github.com/NVIDIA/cutlass/blob/master/examples/48_hopper_warp_specialized_gemm/48_hopper_warp_specialized_gemm.cu) example), I'm getting the following ABI mismatch note related to the [`device_kernel`](https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/device_kernel.h#L103) function:

```
$ nvcc -gencode=arch=compute_90a,code=[sm_90a,compute_90a] -O0 -std=c++17 --expt-relaxed-constexpr cutlass/examples/48_hopper_warp_specialized_gemm/48_hopper_warp_specialized_gemm.cu ...

cutlass/include/cutlass/device_kernel.h: In function ‘void cutlass::device_kernel(typename Operator::Params) [with Operator = cutlass::gemm::kernel::GemmUniversal<cute::tuple<int, int, int>, cutlass::gemm::collective::CollectiveMma<cutlass::gemm::MainloopSm90TmaGmmaWarpSpecialized<7, cute::tuple<cute::constant<int, 1>, cute::constant<int, 2>, cute::constant<int, 1> >, cutlass::gemm::KernelTmaWarpSpecialized>, cute::tuple<cute::constant<int, 128>, cute::constant<int, 128>, cute::constant<int, 32> >, float, cute::tuple<long int, cute::constant<int, 1>, long int>, float, cute::tuple<long int, cute::constant<int, 1>, long int>, cute::TiledMMA<cute::MMA_Atom<cute::SM90_64x128x8_F32TF32TF32_SS_TN<(cute::GMMA::ScaleOut)1, (cute::GMMA::ScaleIn)1, (cute::GMMA::ScaleIn)1> >, cute::Layout<cute::tuple<cute::constant<int, 1>, cute::constant<int, 1>, cute::constant<int, 1> > >, cute::Layout<cute::tuple<cute::constant<int, 1>, cute::constant<int, 1>, cute::constant<int, 1> > >, cute::tuple<cute::Underscore, cute::Underscore, cute::Underscore> >, cute::SM90_TMA_LOAD_MULTICAST, cute::ComposedLayout<cute::Swizzle<3, 4, 3>, cute::smem_ptr_flag_bits<32>, cute::Layout<cute::tuple<cute::constant<int, 8>, cute::constant<int, 32> >, cute::tuple<cute::constant<int, 32>, cute::constant<int, 1> > > >, void, cute::identity, cute::SM90_TMA_LOAD, cute::ComposedLayout<cute::Swizzle<3, 4, 3>, cute::smem_ptr_flag_bits<32>, cute::Layout<cute::tuple<cute::constant<int, 8>, cute::constant<int, 32> >, cute::tuple<cute::constant<int, 32>, cute::constant<int, 1> > > >, void, cute::identity>, cutlass::epilogue::collective::DefaultEpilogue<cute::tuple<cute::constant<int, 1>, long int, long int>, cute::tuple<cute::constant<int, 1>, long int, long int>, cutlass::epilogue::thread::LinearCombination<float, 1, float, float> >, void>]’:
cutlass/include/cutlass/device_kernel.h:103:1: note: The ABI for passing parameters with 64-byte alignment has changed in GCC 4.6
 void device_kernel(CUTLASS_GRID_CONSTANT typename Operator::Params const params)
 ^~~~~~~~~~~~~
```

Is this an issue or something unimportant / to be expected? The compilation succeeds and the resulting binary runs as expected.

Driver version: 525.85.12
CUDA version: 12.0
nvcc version:

```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Fri_Jan__6_16:45:21_PST_2023
Cuda compilation tools, release 12.0, V12.0.140
Build cuda_12.0.r12.0/compiler.32267302_0
```

**Comment by mhoemmen** [comment]
Greetings and thanks for your interest in CUTLASS!

Just to reassure you: GCC's Note here is harmless.  We do not support GCC versions less than 4.6, so the ABI change in GCC 4.6 will not affect customers.  Here is the [GCC issue](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=44948) about this Note.

We investigated ways to suppress the Note.  Here is what we tried.

1. Enclosed the relevant code in a pragma push, `#pragma GCC diagnostic ignored "-Wno-psabi"`, pragma pop sequence.  GCC warned that the option was unknown.
2. Like (1), but with `#pragma GCC diagnostic ignored "-Wpsabi"` instead.
3. Like (1), but with `#pragma GCC diagnostic ignored "-Wall"` instead.
4. Like (1), but with `#pragma GCC diagnostic ignored "-Wabi"` instead.
5. Added `-Wno-psabi` to `CMAKE_CXX_FLAGS`.  (Note that this is an undocumented GCC option. 
 GCC has [an issue open](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=81831) to document it.)
6. Added `-Wno-abi`, `-Wpsabi`, or `-fcompare-debug-second` to `CMAKE_CXX_FLAGS`.

None of these things worked.  We also would not like to suppress notes globally, as it would disable future notes (e.g., ABI changes in more recent compiler versions) that might be more relevant to users.

Some C++ developers suggest working around the issue by passing by const reference (`const Params&`) instead of by value.  However, we cannot do that here, because this is a device kernel.  We need to copy the parameters to device, not give the kernel a reference to host data.  Furthermore, changing a function to pass by const reference instead of by value would break ABI backwards compatibility, thereby causing the issue that the Note warns about.

In summary: Thank you for the report!  This is a known, harmless issue.  We would welcome suggestions for suppressing the Note without hiding other useful Notes and Warnings.

**Comment by aakhundov** [comment]
@mhoemmen Thanks a lot for the detailed clarification! As long as the note is harmless, closing the issue.

**Comment by mhoemmen** [comment]
Glad to help! : - )

**Comment by andyljones** [comment]
While this is immediately harmless, it's swamping our compiler output and making it harder to spot genuine errors - a fix here'd be nice

**Comment by mhoemmen** [comment]
My understanding is that this has been fixed in NVCC itself, by passing along the appropriate flag to GCC.  I can't speak to the release timeline, though.
